{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use hmark_env2\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "import sys\n",
    "from sklearn.linear_model import ElasticNetCV, ElasticNet\n",
    "from sklearn.model_selection import cross_val_score, cross_val_predict, KFold\n",
    "import numpy as np\n",
    "#from xgboost import XGBRegressor\n",
    "from scipy.stats import spearmanr, pearsonr\n",
    "import requests\n",
    "import glob\n",
    "\n",
    "# turn on autoreload\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "source_path = os.path.abspath(os.path.join('..'))\n",
    "if source_path not in sys.path:\n",
    "    sys.path.append(os.path.join(source_path, 'source'))\n",
    "# read source files–––\n",
    "from histone_mark_dataset import histoneMarkDataset\n",
    "from methylation_dataset import methylationDataset\n",
    "import read_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Human/Mouse encode data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'dbxrefs': ['GEO:SAMN40088291'],\n",
       " 'aliases': ['roderic-guigo:td_BE1_H009_biosample'],\n",
       " 'starting_amount_units': 'cells',\n",
       " 'references': [],\n",
       " 'originated_from': {'dbxrefs': [],\n",
       "  'aliases': ['roderic-guigo:td_BE1_biosample'],\n",
       "  'references': [],\n",
       "  'documents': [],\n",
       "  '@type': ['Biosample', 'Item'],\n",
       "  'applied_modifications': [],\n",
       "  'perturbed': False,\n",
       "  'accession': 'ENCBS167WPF',\n",
       "  'source': '/sources/thomas-graf/',\n",
       "  'lab': '/labs/roderic-guigo/',\n",
       "  'characterizations': [],\n",
       "  'uuid': '68c37670-dcfe-4349-9448-6cfcb40884bf',\n",
       "  'treatments': [],\n",
       "  'schema_version': '26',\n",
       "  'award': '/awards/ERC-2011-AdG-294653-RNA-MAPS/',\n",
       "  'donor': '/human-donors/ENCDO305PTK/',\n",
       "  'genetic_modifications': [],\n",
       "  'alternate_accessions': [],\n",
       "  '@id': '/biosamples/ENCBS167WPF/',\n",
       "  'origin_batch': '/biosamples/ENCBS167WPF/',\n",
       "  'summary': 'Homo sapiens BLaER1 cell line',\n",
       "  'organism': '/organisms/human/',\n",
       "  'date_created': '2022-01-27T21:47:19.529052+00:00',\n",
       "  'sex': 'female',\n",
       "  'submitted_by': '/users/15bae39d-9943-4c40-a907-b6f4e542d328/',\n",
       "  'parent_of': [],\n",
       "  'biosample_ontology': {'dbxrefs': ['Cellosaurus:CVCL_VQ57'],\n",
       "   'organ_slims': ['bone element', 'bone marrow'],\n",
       "   'system_slims': ['skeletal system', 'immune system'],\n",
       "   'aliases': ['encode:BLaER1'],\n",
       "   'references': [],\n",
       "   '@type': ['BiosampleType', 'Item'],\n",
       "   'synonyms': ['B cell Leukemia C/EBPalphaER clone 1'],\n",
       "   'term_id': 'EFO:0030018',\n",
       "   'classification': 'cell line',\n",
       "   'uuid': '9843cc9f-627c-470e-abd1-1effc5284b3b',\n",
       "   'schema_version': '1',\n",
       "   'term_name': 'BLaER1',\n",
       "   'cell_slims': [],\n",
       "   'developmental_slims': ['mesoderm'],\n",
       "   'name': 'cell_line_EFO_0030018',\n",
       "   '@id': '/biosample-types/cell_line_EFO_0030018/',\n",
       "   'status': 'released'},\n",
       "  'life_stage': 'unknown',\n",
       "  'internal_tags': [],\n",
       "  'simple_summary': '',\n",
       "  'age': 'unknown',\n",
       "  'status': 'released'},\n",
       " 'documents': [],\n",
       " '@type': ['Biosample', 'Item'],\n",
       " 'applied_modifications': [],\n",
       " 'perturbed': False,\n",
       " 'accession': 'ENCBS902TME',\n",
       " 'source': {'schema_version': '6',\n",
       "  'aliases': [],\n",
       "  '@type': ['Source', 'Item'],\n",
       "  'name': 'thomas-graf',\n",
       "  '@id': '/sources/thomas-graf/',\n",
       "  'title': 'Thomas Graf',\n",
       "  'uuid': '13848833-fa90-4129-8973-f7149230ec40',\n",
       "  'url': 'https://thomas-graf-lab.com/',\n",
       "  'status': 'released'},\n",
       " 'starting_amount': 80000000.0,\n",
       " 'lab': '/labs/roderic-guigo/',\n",
       " 'characterizations': [],\n",
       " 'uuid': 'acc80da4-b93b-4977-9e51-4f1493561c9c',\n",
       " 'treatments': [{'dbxrefs': [],\n",
       "   'amount': 100,\n",
       "   'aliases': [],\n",
       "   'documents': [],\n",
       "   'purpose': 'stimulation',\n",
       "   'date_created': '2021-10-19T18:57:17.530041+00:00',\n",
       "   '@type': ['Treatment', 'Item'],\n",
       "   'treatment_type': 'chemical',\n",
       "   'submitted_by': '/users/43f2f757-5cbf-490a-9787-a1ee85a4cdcd/',\n",
       "   'source': '/sources/calbiochem/',\n",
       "   'amount_units': 'nM',\n",
       "   'treatment_term_id': 'CHEBI:16469',\n",
       "   'uuid': '099aa5e3-9505-4a40-9771-f72e2e5ce484',\n",
       "   'biosamples_used': [],\n",
       "   'schema_version': '12',\n",
       "   'treatment_term_name': '17β-estradiol',\n",
       "   'antibodies_used': [],\n",
       "   'product_id': '3301-1GM',\n",
       "   '@id': '/treatments/099aa5e3-9505-4a40-9771-f72e2e5ce484/',\n",
       "   'status': 'released'},\n",
       "  {'dbxrefs': [],\n",
       "   'amount': 10,\n",
       "   'aliases': [],\n",
       "   'documents': [],\n",
       "   'purpose': 'stimulation',\n",
       "   'date_created': '2021-10-19T18:59:15.079443+00:00',\n",
       "   '@type': ['Treatment', 'Item'],\n",
       "   'treatment_type': 'protein',\n",
       "   'submitted_by': '/users/43f2f757-5cbf-490a-9787-a1ee85a4cdcd/',\n",
       "   'source': '/sources/peprotech/',\n",
       "   'amount_units': 'ng/mL',\n",
       "   'treatment_term_id': 'UniProtKB:P08700',\n",
       "   'uuid': 'b745319b-a1f9-4c54-808b-d943858fd1df',\n",
       "   'biosamples_used': [],\n",
       "   'schema_version': '12',\n",
       "   'treatment_term_name': 'Interleukin-3',\n",
       "   'antibodies_used': [],\n",
       "   'product_id': '200-03',\n",
       "   '@id': '/treatments/b745319b-a1f9-4c54-808b-d943858fd1df/',\n",
       "   'status': 'released'},\n",
       "  {'dbxrefs': [],\n",
       "   'amount': 10,\n",
       "   'aliases': [],\n",
       "   'documents': [],\n",
       "   'purpose': 'stimulation',\n",
       "   'date_created': '2021-10-19T19:01:17.968753+00:00',\n",
       "   '@type': ['Treatment', 'Item'],\n",
       "   'treatment_type': 'protein',\n",
       "   'submitted_by': '/users/43f2f757-5cbf-490a-9787-a1ee85a4cdcd/',\n",
       "   'source': '/sources/peprotech/',\n",
       "   'amount_units': 'ng/mL',\n",
       "   'treatment_term_id': 'UniProtKB:P09603',\n",
       "   'uuid': '467f7c9b-2d09-4242-b1db-266aa0c9154f',\n",
       "   'biosamples_used': [],\n",
       "   'schema_version': '12',\n",
       "   'treatment_term_name': 'CSF1',\n",
       "   'antibodies_used': [],\n",
       "   'product_id': 'AF-300-25',\n",
       "   '@id': '/treatments/467f7c9b-2d09-4242-b1db-266aa0c9154f/',\n",
       "   'status': 'released'}],\n",
       " 'schema_version': '26',\n",
       " 'award': '/awards/ERC-2011-AdG-294653-RNA-MAPS/',\n",
       " 'donor': {'dbxrefs': ['GEO:SAMN40084261'],\n",
       "  'siblings': [],\n",
       "  'aliases': ['encode:BLaER1 donor'],\n",
       "  'notes': 'female with chromosome translocation 1;19 and trisomy 8 (Jack I et al. (1986). Cancer Genet. Cytogenet. 19(3-4):261-269.',\n",
       "  'references': [],\n",
       "  'organism': {'schema_version': '6',\n",
       "   '@type': ['Organism', 'Item'],\n",
       "   'name': 'human',\n",
       "   'taxon_id': '9606',\n",
       "   'scientific_name': 'Homo sapiens',\n",
       "   '@id': '/organisms/human/',\n",
       "   'uuid': '7745b647-ff15-4ff3-9ced-b897d4e2983c',\n",
       "   'status': 'released'},\n",
       "  'documents': [],\n",
       "  'date_created': '2021-12-09T21:56:48.772928+00:00',\n",
       "  '@type': ['HumanDonor', 'Donor', 'Item'],\n",
       "  'sex': 'female',\n",
       "  'submitted_by': '/users/98fb23d3-0d79-4c3d-981d-01539e6589f1/',\n",
       "  'accession': 'ENCDO305PTK',\n",
       "  'lab': '/labs/roderic-guigo/',\n",
       "  'characterizations': [],\n",
       "  'uuid': 'dae48f83-978a-4cb5-8114-7a2ef88b8c38',\n",
       "  'superseded_by': [],\n",
       "  'schema_version': '13',\n",
       "  'award': '/awards/ERC-2011-AdG-294653-RNA-MAPS/',\n",
       "  'children': [],\n",
       "  'alternate_accessions': [],\n",
       "  '@id': '/human-donors/ENCDO305PTK/',\n",
       "  'status': 'released',\n",
       "  'parents': []},\n",
       " 'genetic_modifications': [],\n",
       " 'alternate_accessions': [],\n",
       " '@id': '/biosamples/ENCBS902TME/',\n",
       " 'culture_harvest_date': '2014-05-13',\n",
       " 'origin_batch': '/biosamples/ENCBS902TME/',\n",
       " 'summary': 'Homo sapiens BLaER1 cell line 9 hours after the sample was treated with 100 nM 17β-estradiol, 10 ng/mL Interleukin-3, 10 ng/mL CSF1',\n",
       " 'organism': {'schema_version': '6',\n",
       "  '@type': ['Organism', 'Item'],\n",
       "  'name': 'human',\n",
       "  'taxon_id': '9606',\n",
       "  'scientific_name': 'Homo sapiens',\n",
       "  '@id': '/organisms/human/',\n",
       "  'uuid': '7745b647-ff15-4ff3-9ced-b897d4e2983c',\n",
       "  'status': 'released'},\n",
       " 'date_created': '2022-01-27T21:47:22.697964+00:00',\n",
       " 'sex': 'female',\n",
       " 'submitted_by': {'submits_for': ['/labs/roderic-guigo/'],\n",
       "  '@type': ['User', 'Item'],\n",
       "  '@id': '/users/15bae39d-9943-4c40-a907-b6f4e542d328/',\n",
       "  'lab': '/labs/roderic-guigo/',\n",
       "  'title': 'Emilio Palumbo',\n",
       "  'uuid': '15bae39d-9943-4c40-a907-b6f4e542d328'},\n",
       " 'parent_of': [],\n",
       " 'post_treatment_time_units': 'hour',\n",
       " 'post_treatment_time': '9',\n",
       " 'biosample_ontology': {'dbxrefs': ['Cellosaurus:CVCL_VQ57'],\n",
       "  'organ_slims': ['bone element', 'bone marrow'],\n",
       "  'system_slims': ['skeletal system', 'immune system'],\n",
       "  'aliases': ['encode:BLaER1'],\n",
       "  'references': [],\n",
       "  '@type': ['BiosampleType', 'Item'],\n",
       "  'synonyms': ['B cell Leukemia C/EBPalphaER clone 1'],\n",
       "  'term_id': 'EFO:0030018',\n",
       "  'classification': 'cell line',\n",
       "  'uuid': '9843cc9f-627c-470e-abd1-1effc5284b3b',\n",
       "  'schema_version': '1',\n",
       "  'term_name': 'BLaER1',\n",
       "  'cell_slims': [],\n",
       "  'developmental_slims': ['mesoderm'],\n",
       "  'name': 'cell_line_EFO_0030018',\n",
       "  '@id': '/biosample-types/cell_line_EFO_0030018/',\n",
       "  'status': 'released'},\n",
       " 'life_stage': 'unknown',\n",
       " 'internal_tags': [],\n",
       " 'simple_summary': '9 hours after the sample was treated with 100 nM 17β-estradiol, 10 ng/mL Interleukin-3, 10 ng/mL CSF1',\n",
       " 'culture_start_date': '2014-05-13',\n",
       " 'age': 'unknown',\n",
       " 'status': 'released'}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiments[0]['replicates'][0]['library']['biosample']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_474496/1191992623.py:71: DtypeWarning: Columns (27,58) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  manifest_df = pd.read_csv(manifest_fn, sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "species = \"mouse\"\n",
    "if species == \"human\":\n",
    "    species_query = \"Homo+sapiens\"\n",
    "else:\n",
    "    species_query = \"Mus+musculus\"\n",
    "datatype = \"histone\"\n",
    "\n",
    "# allow classification to be tissue or cell+line\n",
    "if datatype == \"histone\":\n",
    "    url = (\n",
    "        'https://www.encodeproject.org/search/?type=Experiment'\n",
    "        '&status=released'\n",
    "        '&assay_title=Histone+ChIP-seq'   \n",
    "        '&biosample_ontology.classification=tissue'\n",
    "        '&biosample_ontology.classification=cell+line'\n",
    "        f'&replicates.library.biosample.donor.organism.scientific_name={species_query}'\n",
    "        '&frame=embedded'\n",
    "        '&format=json'\n",
    "        '&limit=all'\n",
    "    )\n",
    "else:\n",
    "    url = (\n",
    "        'https://www.encodeproject.org/search/?type=Experiment'\n",
    "    '&status=released'\n",
    "    '&assay_title=WGBS'   \n",
    "    '&assay_title=DNAme+array'   \n",
    "    '&biosample_ontology.classification=tissue'\n",
    "    '&biosample_ontology.classification=cell+line'\n",
    "    f'&replicates.library.biosample.donor.organism.scientific_name={species_query}'\n",
    "    '&frame=embedded'\n",
    "    '&format=json'\n",
    "    '&limit=all'\n",
    "    )\n",
    "\n",
    "r = requests.get(url)\n",
    "experiments = r.json()['@graph']\n",
    "metadata_dict = {}\n",
    "for e in experiments:\n",
    "    # for each replicate\n",
    "    for r in e['replicates']:\n",
    "        sex, age, age_units, donor_accession = np.nan, np.nan, np.nan, np.nan\n",
    "        # get sex\n",
    "        sex = r['library']['biosample']['sex']\n",
    "        # get age\n",
    "        age = r['library']['biosample']['age']\n",
    "        # get age_units\n",
    "        try:\n",
    "            age_units = r['library']['biosample']['age_units']\n",
    "        except(KeyError):\n",
    "            age_units = np.nan\n",
    "        # get accession\n",
    "        accession = r['library']['biosample']['accession']\n",
    "        # get donor accession\n",
    "        donor_accession = r['library']['biosample']['donor']['accession']\n",
    "        # get tissue\n",
    "        tissue = r['library']['biosample']['biosample_ontology']['term_name']\n",
    "        \n",
    "        metadata_dict[accession] = {\n",
    "            'sex': sex, 'age': age, 'age_units': age_units, 'donor_accession': donor_accession, 'tissue': tissue\n",
    "            }\n",
    "        \n",
    "metadata_df = pd.DataFrame(metadata_dict).T\n",
    "metadata_df.set_index('donor_accession', inplace=True)\n",
    "metadata_df = metadata_df.loc[~metadata_df.index.duplicated(keep='first')]\n",
    "\n",
    "download_fn = os.path.join('/cellar/users/zkoch/histone_mark_proj/data/encode', f\"{species}_{datatype}_download.txt\")\n",
    "download_df = pd.read_csv(download_fn, header=None, skiprows=0, sep='\\t').rename(columns={0: 'url'})\n",
    "download_df['File accession'] = download_df['url'].str.split('/').str[4]\n",
    "\n",
    "manifest_fn = os.path.join('/cellar/users/zkoch/histone_mark_proj/data/encode/metadata', f\"{species}_{datatype}_metadata_download.tsv\")\n",
    "manifest_df = pd.read_csv(manifest_fn, sep='\\t')\n",
    "manifest_df['donor'] = manifest_df['Donor(s)'].str.split('/').str[2]\n",
    "\n",
    "downloaded_files = download_df['File accession'].unique()\n",
    "manifest_of_downloaded_files = manifest_df.query(\"`File accession` in @downloaded_files\")\n",
    "manifest_of_downloaded_files.reset_index(drop=True, inplace=True)\n",
    "manifest_of_downloaded_files = manifest_of_downloaded_files[['File accession', 'Biosample term name', 'Biosample type', 'Experiment target', 'donor']]\n",
    "manifest_of_downloaded_files.rename(columns = {'Biosample term name': 'tissue', 'Biosample type':'sample_type', 'Experiment target': 'histone_mark'}, inplace=True)\n",
    "\n",
    "manifest_of_downloaded_files['age'] = manifest_of_downloaded_files['donor'].map(metadata_df['age'])\n",
    "manifest_of_downloaded_files['age_units'] = manifest_of_downloaded_files['donor'].map(metadata_df['age_units'])\n",
    "manifest_of_downloaded_files['sex'] = manifest_of_downloaded_files['donor'].map(metadata_df['sex'])\n",
    "\n",
    "# convert age to float, dropping nan first\n",
    "manifest_of_downloaded_files.query(\"age != 'unknown'\", inplace=True)\n",
    "manifest_of_downloaded_files['age'] = manifest_of_downloaded_files['age'].replace('90 or above', '90').replace('2-4', np.nan)\n",
    "manifest_of_downloaded_files['age'] = manifest_of_downloaded_files['age'].astype(float)\n",
    "metadata_df.dropna(subset=['age'], inplace=True)\n",
    "\n",
    "# for ages with units of weeks, divide by 52\n",
    "manifest_of_downloaded_files['age_years'] = manifest_of_downloaded_files['age']\n",
    "manifest_of_downloaded_files.loc[manifest_of_downloaded_files['age_units'] == 'week', 'age_years'] = manifest_of_downloaded_files.loc[manifest_of_downloaded_files['age_units'] == 'week', 'age_years'] / 52\n",
    "# for ages with units of days, divide by 365\n",
    "manifest_of_downloaded_files.loc[manifest_of_downloaded_files['age_units'] == 'day', 'age_years'] = manifest_of_downloaded_files.loc[manifest_of_downloaded_files['age_units'] == 'day', 'age_years'] / 365\n",
    "# for ages with units of months, divide by 12\n",
    "manifest_of_downloaded_files.loc[manifest_of_downloaded_files['age_units'] == 'month', 'age_years'] = manifest_of_downloaded_files.loc[manifest_of_downloaded_files['age_units'] == 'month', 'age_years'] / 12\n",
    "# drop age_units\n",
    "manifest_of_downloaded_files.drop(columns=['age_units', 'age'], inplace=True)\n",
    "# create histone column\n",
    "if datatype == \"histone\":\n",
    "    manifest_of_downloaded_files['histone_mark'] = manifest_of_downloaded_files['histone_mark'].str.split('-').str[0]\n",
    "else:\n",
    "    # drop histone_mark column\n",
    "    manifest_of_downloaded_files.drop(columns=['histone_mark'], inplace=True)\n",
    "\n",
    "manifest_of_downloaded_files.to_parquet(os.path.join('/cellar/users/zkoch/histone_mark_proj/data/encode/metadata', f\"{species}_{datatype}_metadata_fixed.parquet\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File accession</th>\n",
       "      <th>tissue</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>histone_mark</th>\n",
       "      <th>donor</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_years</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCFF823VWF</td>\n",
       "      <td>heart</td>\n",
       "      <td>tissue</td>\n",
       "      <td>H3K4me3</td>\n",
       "      <td>ENCDO956IXV</td>\n",
       "      <td>mixed</td>\n",
       "      <td>0.039726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCFF551GXY</td>\n",
       "      <td>erythroblast</td>\n",
       "      <td>primary cell</td>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>ENCDO083AAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCFF861SPX</td>\n",
       "      <td>G1E</td>\n",
       "      <td>cell line</td>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>ENCDO089AAA</td>\n",
       "      <td>male</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENCFF682TYN</td>\n",
       "      <td>embryonic fibroblast</td>\n",
       "      <td>primary cell</td>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>ENCDO072AAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENCFF728FZO</td>\n",
       "      <td>erythroblast</td>\n",
       "      <td>primary cell</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>ENCDO083AAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>ENCFF858DIX</td>\n",
       "      <td>liver</td>\n",
       "      <td>tissue</td>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>ENCDO469XMJ</td>\n",
       "      <td>mixed</td>\n",
       "      <td>0.039726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>ENCFF188ILR</td>\n",
       "      <td>embryonic facial prominence</td>\n",
       "      <td>tissue</td>\n",
       "      <td>H3K36me3</td>\n",
       "      <td>ENCDO469XMJ</td>\n",
       "      <td>mixed</td>\n",
       "      <td>0.039726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>ENCFF174MOQ</td>\n",
       "      <td>lung</td>\n",
       "      <td>tissue</td>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>ENCDO956IXV</td>\n",
       "      <td>mixed</td>\n",
       "      <td>0.039726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>653</th>\n",
       "      <td>ENCFF736SSJ</td>\n",
       "      <td>limb</td>\n",
       "      <td>tissue</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>ENCDO956IXV</td>\n",
       "      <td>mixed</td>\n",
       "      <td>0.039726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>ENCFF247CKJ</td>\n",
       "      <td>forebrain</td>\n",
       "      <td>tissue</td>\n",
       "      <td>H3K36me3</td>\n",
       "      <td>ENCDO956IXV</td>\n",
       "      <td>mixed</td>\n",
       "      <td>0.039726</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>598 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    File accession                       tissue   sample_type histone_mark  \\\n",
       "0      ENCFF823VWF                        heart        tissue      H3K4me3   \n",
       "2      ENCFF551GXY                 erythroblast  primary cell      H3K9me3   \n",
       "3      ENCFF861SPX                          G1E     cell line      H3K9me3   \n",
       "5      ENCFF682TYN         embryonic fibroblast  primary cell      H3K27ac   \n",
       "6      ENCFF728FZO                 erythroblast  primary cell     H3K27me3   \n",
       "..             ...                          ...           ...          ...   \n",
       "650    ENCFF858DIX                        liver        tissue      H3K9me3   \n",
       "651    ENCFF188ILR  embryonic facial prominence        tissue     H3K36me3   \n",
       "652    ENCFF174MOQ                         lung        tissue      H3K9me3   \n",
       "653    ENCFF736SSJ                         limb        tissue     H3K27me3   \n",
       "654    ENCFF247CKJ                    forebrain        tissue     H3K36me3   \n",
       "\n",
       "           donor    sex  age_years  \n",
       "0    ENCDO956IXV  mixed   0.039726  \n",
       "2    ENCDO083AAA    NaN        NaN  \n",
       "3    ENCDO089AAA   male   0.000000  \n",
       "5    ENCDO072AAA    NaN        NaN  \n",
       "6    ENCDO083AAA    NaN        NaN  \n",
       "..           ...    ...        ...  \n",
       "650  ENCDO469XMJ  mixed   0.039726  \n",
       "651  ENCDO469XMJ  mixed   0.039726  \n",
       "652  ENCDO956IXV  mixed   0.039726  \n",
       "653  ENCDO956IXV  mixed   0.039726  \n",
       "654  ENCDO956IXV  mixed   0.039726  \n",
       "\n",
       "[598 rows x 7 columns]"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "manifest_of_downloaded_files#['donor'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Blueprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "blueprint_metadata_fn = \"/cellar/users/zkoch/histone_mark_proj/data/blueprint/blueprint_file_metadata.tsv\"\n",
    "blueprint_metadata_df = pd.read_csv(blueprint_metadata_fn, sep=\"\\t\")\n",
    "interesting_cols = [\n",
    "    'SAMPLE_ID', # ERS954093\n",
    "    'SAMPLE_NAME', # S00XDKU1\n",
    "    'DONOR_ID', # 15548, WR27 # same as SAMPLE_BARCODE\n",
    "    'DONOR_AGE', # 55 - 60\n",
    "    'DONOR_SEX', # Male, Female\n",
    "    'SAMPLE_SOURCE', # cell line\n",
    "    'EXPERIMENT_TYPE',\n",
    "    'DISEASE',\n",
    "    'BIOMATERIAL_TYPE',\n",
    "    'TISSUE_TYPE',\n",
    "    'FILE', # 'blueprint/data/homo_sapiens/GRCh38/bone_marrow/15548/Multiple_Myeloma/Bisulfite-Seq/CNAG/S00XDKU1.CPG_methylation_calls.bs_call.GRCh38.20160531.bw'\n",
    "    'FILE_TYPE'\n",
    "    ]\n",
    "blueprint_metadata_df = blueprint_metadata_df[interesting_cols]\n",
    "\n",
    "exp_types = ['DNA Methylation', 'H3K27ac','H3K4me3', 'H3K4me1', 'H3K36me3', 'H3K9me3', 'H3K27me3' ]\n",
    "blueprint_metadata_df.query(\"EXPERIMENT_TYPE in @exp_types\", inplace=True)\n",
    "\n",
    "# get rid of cell lines\n",
    "blueprint_metadata_df.query(\"SAMPLE_SOURCE != 'Cell Line'\", inplace=True)\n",
    "\n",
    "# get samples with a non nan age\n",
    "blueprint_metadata_df = blueprint_metadata_df.loc[blueprint_metadata_df['DONOR_AGE'].notna()]\n",
    "\n",
    "# fill nan disease with 'Healthy'\n",
    "blueprint_metadata_df['DISEASE'] = blueprint_metadata_df['DISEASE'].fillna('Healthy')\n",
    "\n",
    "# BS_METH_CALL_CNAG is bw, but can convert to bed\n",
    "file_types = ['CHIP_MACS2_BROAD_BED', 'CHIP_MACS2_BED', 'BS_METH_CALL_CNAG']\n",
    "blueprint_metadata_df.query(\"FILE_TYPE in @file_types\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of donors with DNA Methylation and histone marks:  100\n"
     ]
    }
   ],
   "source": [
    "# find donors with DNA Methylation and at least 2 experiments\n",
    "donors_w_dnam = blueprint_metadata_df.query(\"EXPERIMENT_TYPE == 'DNA Methylation'\")['DONOR_ID'].unique()\n",
    "donors_w_histone = blueprint_metadata_df.query(\"EXPERIMENT_TYPE in @exp_types and EXPERIMENT_TYPE != 'DNA Methylation'\")['DONOR_ID'].unique()\n",
    "print(\"Number of donors with DNA Methylation and histone marks: \", len(set(donors_w_histone).intersection(set(donors_w_dnam))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HistoneMarkDataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading narrowPeak files: 100%|██████████| 2515/2515 [00:28<00:00, 88.04it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading regulatory features files\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Creating data matrix: 100%|██████████| 2515/2515 [04:48<00:00,  8.71it/s]\n"
     ]
    }
   ],
   "source": [
    "# 12 min for all at 16 cores\n",
    "human_encode_hmark = histoneMarkDataset(\n",
    "    name = \"human_encode_hmark\",\n",
    "    species = \"homo_sapiens\",\n",
    "    narrowpeak_dir = \"/cellar/users/zkoch/histone_mark_proj/data/encode/human_bed_files/histone\",\n",
    "    regulatory_features_path = \"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/Homo_sapiens.GRCh38.regulatory_features.v112.gff3.gz\",\n",
    "    gene_locations_path = \"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/Homo_sapiens.GRCh38.112.gtf.gz\", \n",
    "    metadata_fn = os.path.join('/cellar/users/zkoch/histone_mark_proj/data/encode/metadata', f\"human_histone_metadata_final.parquet\"),\n",
    "    n_cores = 16\n",
    ")\n",
    "human_encode_hmark.create_data_matrix()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading narrowPeak files: 100%|██████████| 655/655 [00:00<00:00, 202810.36it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading regulatory features files\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Creating data matrix: 100%|██████████| 655/655 [01:09<00:00,  9.38it/s]\n"
     ]
    }
   ],
   "source": [
    "# 3 min for all at 16 cores\n",
    "mouse_encode_hmark = histoneMarkDataset(\n",
    "    name = \"mouse_encode_hmark\",\n",
    "    species = \"mus_musculus\",\n",
    "    narrowpeak_dir = \"/cellar/users/zkoch/histone_mark_proj/data/encode/mouse_bed_files2/histone\",\n",
    "    regulatory_features_path = \"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/mus_musculus.GRCm38.Regulatory_Build.regulatory_features.20180516.gff.gz\",\n",
    "    gene_locations_path = \"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/Mus_musculus.GRCm38.102.gtf.gz\", \n",
    "    n_cores = 16\n",
    ")\n",
    "mouse_encode_hmark.create_data_matrix()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_hist_metadata_df = pd.read_csv(os.path.join('/cellar/users/zkoch/histone_mark_proj/data/encode/metadata', \"human_metadata.csv\")).rename(columns = {'Unnamed: 0': 'File accession'})\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Wang 2017\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Available keys: ['/covariates', '/methy_mat']\n"
     ]
    }
   ],
   "source": [
    "# Method 1: Using pandas\n",
    "with pd.HDFStore('/cellar/users/zkoch/histone_mark_proj/data/wang_2017/data/AllData_combatnonpara_logage.h5', 'r') as store:\n",
    "    print(\"Available keys:\", store.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "cov_df = pd.read_hdf(\"/cellar/users/zkoch/histone_mark_proj/data/wang_2017/data/AllData_non_normalized_data_logage.h5\", key = 'covariates')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age_days</th>\n",
       "      <th>count</th>\n",
       "      <th>unlogged_age_days</th>\n",
       "      <th>age_months</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.807355</td>\n",
       "      <td>10</td>\n",
       "      <td>7.00</td>\n",
       "      <td>0.225806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4.392317</td>\n",
       "      <td>9</td>\n",
       "      <td>21.00</td>\n",
       "      <td>0.677419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5.807355</td>\n",
       "      <td>10</td>\n",
       "      <td>56.00</td>\n",
       "      <td>1.806452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5.926948</td>\n",
       "      <td>12</td>\n",
       "      <td>60.84</td>\n",
       "      <td>1.962581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.977280</td>\n",
       "      <td>38</td>\n",
       "      <td>63.00</td>\n",
       "      <td>2.032258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6.926948</td>\n",
       "      <td>2</td>\n",
       "      <td>121.68</td>\n",
       "      <td>3.925161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7.129283</td>\n",
       "      <td>51</td>\n",
       "      <td>140.00</td>\n",
       "      <td>4.516129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7.614710</td>\n",
       "      <td>15</td>\n",
       "      <td>196.00</td>\n",
       "      <td>6.322581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7.761551</td>\n",
       "      <td>4</td>\n",
       "      <td>217.00</td>\n",
       "      <td>7.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9.386380</td>\n",
       "      <td>20</td>\n",
       "      <td>669.24</td>\n",
       "      <td>21.588387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9.627388</td>\n",
       "      <td>2</td>\n",
       "      <td>790.92</td>\n",
       "      <td>25.513548</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    age_days  count  unlogged_age_days  age_months\n",
       "0   2.807355     10               7.00    0.225806\n",
       "1   4.392317      9              21.00    0.677419\n",
       "2   5.807355     10              56.00    1.806452\n",
       "3   5.926948     12              60.84    1.962581\n",
       "4   5.977280     38              63.00    2.032258\n",
       "5   6.926948      2             121.68    3.925161\n",
       "6   7.129283     51             140.00    4.516129\n",
       "7   7.614710     15             196.00    6.322581\n",
       "8   7.761551      4             217.00    7.000000\n",
       "9   9.386380     20             669.24   21.588387\n",
       "10  9.627388      2             790.92   25.513548"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = cov_df['age_days'].value_counts().sort_index().reset_index()\n",
    "df['unlogged_age_days'] = 2**df['age_days']\n",
    "df['age_months'] = df['unlogged_age_days'] / 31\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in from .h5\n",
    "methyl_df = pd.read_hdf(\"/cellar/users/zkoch/histone_mark_proj/data/wang_2017/data/AllData_non_normalized_data_logage.h5\", key = 'methy_mat')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hillje"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20-Dec-2024 11:16:00 DEBUG utils - Directory /cellar/users/zkoch/histone_mark_proj/data/hillje_2022 already exists. Skipping.\n",
      "20-Dec-2024 11:16:00 INFO GEOparse - Downloading ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE179nnn/GSE179090/soft/GSE179090_family.soft.gz to /cellar/users/zkoch/histone_mark_proj/data/hillje_2022/GSE179090_family.soft.gz\n",
      "100%|██████████| 7.69k/7.69k [00:00<00:00, 23.6kB/s]\n",
      "20-Dec-2024 11:16:01 DEBUG downloader - Size validation passed\n",
      "20-Dec-2024 11:16:01 DEBUG downloader - Moving /tmp/tmp_95cneib to /cellar/users/zkoch/histone_mark_proj/data/hillje_2022/GSE179090_family.soft.gz\n",
      "20-Dec-2024 11:16:01 DEBUG downloader - Successfully downloaded ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE179nnn/GSE179090/soft/GSE179090_family.soft.gz\n",
      "20-Dec-2024 11:16:01 INFO GEOparse - Parsing /cellar/users/zkoch/histone_mark_proj/data/hillje_2022/GSE179090_family.soft.gz: \n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SERIES: GSE179090\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - PLATFORM: GPL13112\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406230\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406231\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406232\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406233\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406234\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406235\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406236\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406237\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406238\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406239\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406240\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406241\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406242\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406243\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406244\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406245\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406246\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406247\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406248\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406249\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406250\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406251\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406252\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406253\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406254\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406255\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406256\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406257\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406258\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406259\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406260\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406261\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406262\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406263\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406264\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406265\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406266\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406267\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406268\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406269\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406270\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406271\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406272\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406273\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406274\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406275\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406276\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406277\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406278\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406279\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406280\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406281\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406282\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406283\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406284\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406285\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406286\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406287\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406288\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406289\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406290\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406291\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406292\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406293\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406294\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406295\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406296\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406297\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406298\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406299\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406300\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406301\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406302\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406303\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406304\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406305\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406306\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406307\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406308\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406309\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406310\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406311\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406312\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406313\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406314\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406315\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406316\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406317\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406318\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406319\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406320\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406321\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406322\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406323\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406324\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406325\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406326\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406327\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406328\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406329\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406330\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406331\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406332\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406333\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406334\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406335\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406336\n",
      "20-Dec-2024 11:16:01 DEBUG GEOparse - SAMPLE: GSM5406337\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import os\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "import GEOparse\n",
    "\n",
    "gse = GEOparse.get_GEO(geo=\"GSE179090\",\n",
    "                      destdir=\"/cellar/users/zkoch/histone_mark_proj/data/hillje_2022\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df = pd.read_csv(\"/cellar/users/zkoch/histone_mark_proj/data/hillje_2022/SraRunTable.csv\")\n",
    "metadata_df.rename(columns = {'Run': 'donor', 'AGE': 'age_months', 'chip_antibody': 'histone_mark', 'treatment':'diet'}, inplace=True)\n",
    "metadata_df['age_years'] = metadata_df['age_months'].str.replace(' months', '').astype(float) / 52\n",
    "metadata_df.set_index('donor', inplace=True)\n",
    "# remove index name\n",
    "metadata_df.index.name = None\n",
    "metadata_df['donor'] = metadata_df.index\n",
    "metadata_df = metadata_df[['donor','age_years', 'histone_mark', 'tissue', 'diet']]\n",
    "metadata_df.to_csv(\"/cellar/users/zkoch/histone_mark_proj/data/hillje_2022/metadata.csv\", index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading from parquet files\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading narrowPeak files: 100%|██████████| 108/108 [00:00<00:00, 95045.08it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading regulatory features files\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Creating data matrix: 100%|██████████| 108/108 [00:07<00:00, 13.61it/s]\n",
      "Populating data matrix: 100%|██████████| 108/108 [00:01<00:00, 88.82it/s]\n"
     ]
    }
   ],
   "source": [
    "# 12 min for all at 16 cores\n",
    "hillje = histoneMarkDataset(\n",
    "    name = \"hillje\",\n",
    "    species = \"mouse\",\n",
    "    narrowpeak_dir = \"/cellar/users/zkoch/histone_mark_proj/data/hillje_2022/broad_peaks\",\n",
    "    regulatory_features_path = \"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/mus_musculus.GRCm38.Regulatory_Build.regulatory_features.20180516.gff.gz\",\n",
    "    gene_locations_path = \"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/Mus_musculus.GRCm38.102.gtf.gz\", \n",
    "    metadata_df = pd.read_csv(\"/cellar/users/zkoch/histone_mark_proj/data/hillje_2022/metadata.csv\", index_col=0),\n",
    "    n_cores = 16,\n",
    "    split_on = '_',\n",
    "    overwrite_cached_data_matrix = True,\n",
    "    use_cached_data_matrix = False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Stubbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "stubbs_dir = \"/cellar/users/zkoch/histone_mark_proj/data/stubbs_2017\"\n",
    "metadata_df = pd.read_csv(os.path.join(stubbs_dir, \"metadata.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metadata preview:\n",
      "                 id  age_weeks  tissue\n",
      "0  M02NB_1wk_Cortex          1  Cortex\n",
      "1  M01NB_1wk_Cortex          1  Cortex\n",
      "2  M03NB_1wk_Cortex          1  Cortex\n",
      "3  M04NB_1wk_Cortex          1  Cortex\n",
      "4   M01NB_1wk_Heart          1   Heart\n",
      "\n",
      "Saved to: /cellar/users/zkoch/histone_mark_proj/data/stubbs_2017/metadata.csv\n"
     ]
    }
   ],
   "source": [
    "# Create metadata DataFrame from filenames\n",
    "files = glob.glob(os.path.join(stubbs_dir, \"*.bed.gz\"))\n",
    "metadata = []\n",
    "\n",
    "for f in files:\n",
    "    basename = os.path.basename(f)\n",
    "    # Split filename to get components\n",
    "    parts = basename.replace(\".bed.gz\", \"\").split(\"_\")\n",
    "    \n",
    "    # Extract metadata\n",
    "    sample_id = parts[0]\n",
    "    age = int(parts[1].replace(\"wk\", \"\"))  # Remove 'wk' and convert to int\n",
    "    tissue = parts[2]\n",
    "    \n",
    "    metadata.append({\n",
    "        'id': basename.replace(\".bed.gz\", \"\"),\n",
    "        'age_weeks': age,\n",
    "        'tissue': tissue\n",
    "    })\n",
    "# Create DataFrame and sort\n",
    "metadata_df = pd.DataFrame(metadata)\n",
    "metadata_df = metadata_df.sort_values(['age_weeks', 'tissue']).reset_index(drop=True)\n",
    "\n",
    "# Save metadata\n",
    "metadata_path = os.path.join(stubbs_dir, \"metadata.csv\")\n",
    "metadata_df.to_csv(metadata_path, index=False)\n",
    "\n",
    "print(\"Metadata preview:\")\n",
    "print(metadata_df.head())\n",
    "print(\"\\nSaved to:\", metadata_path)\n",
    "\n",
    "metadata_df['age_years'] = metadata_df['age_weeks'] / 52\n",
    "metadata_df.rename(columns = {'id': 'donor'}, inplace=True)\n",
    "metadata_df.drop(columns=['age_weeks'], inplace=True)\n",
    "metadata_df.set_index('donor', inplace=True)\n",
    "metadata_df['donor'] = metadata_df.index\n",
    "metadata_df.to_csv(os.path.join(stubbs_dir, \"metadata.csv\"), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize methylation dataset\n",
    "stubbs_methyl = methylationDataset(\n",
    "    name=\"stubbs_methyl\",\n",
    "    species=\"mouse\",\n",
    "    bedmethyl_dir=stubbs_dir,\n",
    "    regulatory_features_path=\"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/mus_musculus.GRCm38.Regulatory_Build.regulatory_features.20180516.gff.gz\",\n",
    "    gene_locations_path=\"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/Mus_musculus.GRCm38.102.gtf.gz\",\n",
    "    metadata_df=metadata_df,\n",
    "    n_cores=20,\n",
    "    use_cached_data_matrix=False,\n",
    "    overwrite_cached_data_matrix=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_parquet('/cellar/users/zkoch/histone_mark_proj/data/stubbs_2017/M00018754_27wk_Lung_mean_signal.parquet')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CEEHRC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"/cellar/users/zkoch/histone_mark_proj/data/CEEHRC/CEERHC_metadata.csv\")\n",
    "df = df[['id', 'cell_type', 'donor_age', 'donor_age_unit', 'tissue_type']]\n",
    "df.rename(columns = {'id': 'donor', 'tissue_type': 'tissue', 'donor_age': 'age_years'}, inplace = True)\n",
    "# drop rows where age_years is not 'year'\n",
    "df = df[df['donor_age_unit'] == 'year']\n",
    "# drop rows where age_years is == 90+\n",
    "df = df[df['age_years'] != '90+']\n",
    "# convert rows where age_years has '-' or ' - ' to be the average of the two values on either side of the dash\n",
    "# for other rows just convert to float\n",
    "df['age_years'] = df['age_years'].apply(lambda x: float(x) if not isinstance(x, str) or '-' not in x else float(sum([float(i.strip()) for i in x.replace(' ', '').split('-')])/2))\n",
    "# combine cell_type and tissue to make a tissue column, using tissue where both are available, otherwise use cell_type\n",
    "df['tissue'] = df.apply(lambda row: row['tissue'] if pd.notna(row['tissue']) else row['cell_type'], axis=1)\n",
    "# drop cell_type column\n",
    "df.drop(columns = ['cell_type'], inplace = True)\n",
    "# drop donor_age_unit column\n",
    "df.drop(columns = ['donor_age_unit'], inplace = True)\n",
    "# drop nans\n",
    "df.dropna(inplace = True)\n",
    "\n",
    "# Get all parquet files in the data directories\n",
    "import glob\n",
    "import os\n",
    "\n",
    "methyl_parquet_files = glob.glob('/cellar/users/zkoch/histone_mark_proj/data/CEEHRC/methyl_bed/*methylation_profile.parquet')\n",
    "histone_parquet_files = glob.glob('/cellar/users/zkoch/histone_mark_proj/data/CEEHRC/bed_peaks/*.parquet')\n",
    "\n",
    "# Create lists to store the parsed values\n",
    "samples = []\n",
    "donors = []\n",
    "marks = []\n",
    "\n",
    "# Parse each filename\n",
    "for f in methyl_parquet_files:\n",
    "    parts = os.path.basename(f).split('.')\n",
    "    samples.append(parts[0])\n",
    "    donors.append(parts[2]) \n",
    "    marks.append('methylation')\n",
    "for f in histone_parquet_files:\n",
    "    parts = os.path.basename(f).split('.')\n",
    "    samples.append(parts[0])\n",
    "    donors.append(parts[2])\n",
    "    marks.append(parts[3])\n",
    "\n",
    "# Create dataframe\n",
    "files_df = pd.DataFrame({\n",
    "    'sample': samples,\n",
    "    'donor': donors,\n",
    "    'mark': marks\n",
    "})\n",
    "# merge with metadata on donor\n",
    "new_metadata_df = files_df.merge(df, on = 'donor', how = 'outer').dropna()\n",
    "new_metadata_df.set_index('sample', inplace = True)\n",
    "new_metadata_df.to_csv(\"/cellar/users/zkoch/histone_mark_proj/data/CEEHRC/metadata_final.csv\", index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading cached data matrix from /cellar/users/zkoch/histone_mark_proj/data/CEEHRC/bed_peaks/data_matrix/ceehrc_hmark_data_matrix_w_meta.parquet\n"
     ]
    }
   ],
   "source": [
    "ceehrc_hmark = histoneMarkDataset(\n",
    "            name = \"ceehrc_hmark\",\n",
    "            species = \"human\",\n",
    "            narrowpeak_dir = \"/cellar/users/zkoch/histone_mark_proj/data/CEEHRC/bed_peaks\",\n",
    "            regulatory_features_path = \"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/Homo_sapiens.GRCh38.regulatory_features.v112.gff3.gz\",\n",
    "            gene_locations_path = \"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/Homo_sapiens.GRCh38.112.gtf.gz\",\n",
    "            sample_name_part = 2,\n",
    "            split_on = '.',\n",
    "            metadata_df = metadata_df,\n",
    "            n_cores = 8,\n",
    "            use_cached_data_matrix = True,\n",
    "            overwrite_cached_data_matrix = False,\n",
    "            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df = pd.read_csv(\"/cellar/users/zkoch/histone_mark_proj/data/CEEHRC/metadata_final.csv\", index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ceehrc_methyl = methylationDataset(\n",
    "    name = \"ceehrc_methyl\",\n",
    "    species = \"human\",\n",
    "    bedmethyl_dir = \"/cellar/users/zkoch/histone_mark_proj/data/CEEHRC/methyl_bed\",\n",
    "    regulatory_features_path = \"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/Homo_sapiens.GRCh38.regulatory_features.v112.gff3.gz\",\n",
    "    gene_locations_path = \"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/Homo_sapiens.GRCh38.112.gtf.gz\",\n",
    "    metadata_df = metadata_df,\n",
    "    n_cores = 12,\n",
    "    use_cached_data_matrix = False,\n",
    "    overwrite_cached_data_matrix = True,\n",
    "    sample_name_part = 2,\n",
    "    split_on = '.',\n",
    "    start_sample_number = 0,\n",
    "    end_sample_number = 2\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Petkovich"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df = pd.read_csv(\"/cellar/users/zkoch/histone_mark_proj/data/petkovich_2017/GSE80672_series_matrix.txt.gz\", skiprows=34, sep='\\t', header=None)\n",
    "# transpose metadata_df\n",
    "metadata_df = metadata_df.T\n",
    "# set first row as header\n",
    "metadata_df.columns = metadata_df.iloc[0]\n",
    "# drop first row\n",
    "metadata_df = metadata_df.iloc[1:]\n",
    "\n",
    "metadata_df = metadata_df[['!Sample_geo_accession', '!Sample_source_name_ch1', '!Sample_characteristics_ch1' ]]\n",
    "metadata_df.columns = ['donor', 'tissue', 'tissue2', 'strain2', 'age_months', 'sex', 'strain', 'diet']\n",
    "metadata_df.drop(columns = ['tissue2', 'strain2'], inplace = True)\n",
    "metadata_df['age_years'] = metadata_df['age_months'].str.split(':').str[-1].str.split('(').str[0].str.strip().astype(float) / 12\n",
    "metadata_df.drop(columns = ['age_months', 'sex'], inplace = True)\n",
    "metadata_df['strain'] = metadata_df['strain'].str.split(': ').str[-1].str.strip()\n",
    "metadata_df['diet'] = metadata_df['diet'].str.split(':').str[-1].str.strip().map({'standard': 'standard', 'Calorie Restricted': 'CR'})\n",
    "metadata_df.set_index('donor', inplace = True)\n",
    "metadata_df['donor'] = metadata_df.index\n",
    "metadata_df.index.name = None\n",
    "metadata_df.to_csv(\"/cellar/users/zkoch/histone_mark_proj/data/petkovich_2017/metadata_final.csv\", index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df = pd.read_csv(\"/cellar/users/zkoch/histone_mark_proj/data/petkovich_2017/metadata_final.csv\", index_col=0)\n",
    "petkovich_methyl = methylationDataset(\n",
    "    name = \"petkovich\",\n",
    "    species = \"mouse\",\n",
    "    bedmethyl_dir = \"/cellar/users/zkoch/histone_mark_proj/data/petkovich_2017/bed_methyl\",\n",
    "    regulatory_features_path = \"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/mus_musculus.GRCm38.Regulatory_Build.regulatory_features.20180516.gff.gz\",\n",
    "    gene_locations_path = \"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/Mus_musculus.GRCm38.102.gtf.gz\",\n",
    "    metadata_df = metadata_df,\n",
    "    n_cores = 18,\n",
    "    use_cached_data_matrix = False,\n",
    "    overwrite_cached_data_matrix = True,\n",
    "    sample_name_part = 0,\n",
    "    split_on = '_',\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading dataset: petkovich\n",
      "Reading cached data matrix from /cellar/users/zkoch/histone_mark_proj/data/petkovich_2017/bed_methyl/data_matrix/petkovich_data_matrix_w_meta.parquet\n"
     ]
    }
   ],
   "source": [
    "loader = read_data.DatasetLoader(\n",
    "    dataset_name = \"petkovich\",\n",
    ")\n",
    "petkovich = loader.load_dataset()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Meer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df = pd.read_csv(\n",
    "                    '/cellar/users/zkoch/histone_mark_proj/data/meer_2018/GSE121141_series_matrix.txt',\n",
    "                    skiprows=34,\n",
    "                    sep='\\t',\n",
    "                    header=None\n",
    "                    ).T\n",
    "metadata_df.columns = metadata_df.iloc[0]\n",
    "metadata_df = metadata_df.iloc[1:]\n",
    "metadata_df = metadata_df[[\"ID_REF\", \"!Sample_characteristics_ch1\"]]\n",
    "metadata_df.columns = ['donor', 'strain2', 'sex', 'age_months', 'strain', 'tissue']\n",
    "metadata_df.drop(columns = ['strain2', 'sex',  'strain',], inplace = True)\n",
    "metadata_df['age_years'] = metadata_df['age_months'].str.split(':').str[-1].str.strip().astype(float) / 12\n",
    "metadata_df.drop(columns = ['age_months'], inplace = True)\n",
    "metadata_df['tissue'] = metadata_df['tissue'].str.split(':').str[-1].str.strip()\n",
    "metadata_df.set_index('donor', inplace = True)\n",
    "metadata_df['donor'] = metadata_df.index\n",
    "metadata_df.index.name = None\n",
    "metadata_df.to_csv(\"/cellar/users/zkoch/histone_mark_proj/data/meer_2018/metadata_final.csv\", index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df = pd.read_csv(\"/cellar/users/zkoch/histone_mark_proj/data/meer_2018/metadata_final.csv\", index_col=0)\n",
    "# Initialize methylation dataset\n",
    "meer = methylationDataset(\n",
    "    name=\"meer\",\n",
    "    species=\"mouse\",\n",
    "    bedmethyl_dir=\"/cellar/users/zkoch/histone_mark_proj/data/meer_2018/bed_methyl\",\n",
    "    regulatory_features_path=\"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/mus_musculus.GRCm38.Regulatory_Build.regulatory_features.20180516.gff.gz\",\n",
    "    gene_locations_path=\"/cellar/users/zkoch/histone_mark_proj/data/ensembl_reference/Mus_musculus.GRCm38.102.gtf.gz\",\n",
    "    metadata_df=metadata_df,\n",
    "    n_cores=20,\n",
    "    use_cached_data_matrix=False,\n",
    "    overwrite_cached_data_matrix=True,\n",
    "    split_on = '_',\n",
    "    sample_name_part = 0,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hmark_env2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
